cdap-watchdog/src/main/java/io/cdap/cdap/logging/appender/CompositeLogAppender.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright © 2024 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.logging.appender;
+
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.Context;
+import io.cdap.cdap.logging.framework.local.LocalLogAppender;
+import java.util.List;
+import org.slf4j.LoggerFactory;
+
+public class CompositeLogAppender extends LogAppender {
+
+  private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(CompositeLogAppender.class);
+  private final List<LogAppender> appenders;
+
+  public CompositeLogAppender(List<LogAppender> appenders) {
+    this.appenders = appenders;
+    setName(getClass().getName());
+  }
+
+  @Override
+  public void start() {
+    synchronized (appenders) {
+      appenders.forEach(appender -> safelyExecute(
+          appender::start,
+          String.format("Failed to start appender: %s", appender.getName())
+      ));
+    }
+    super.start();
+  }
+
+  @Override
+  public void stop() {
+    super.stop();
+    synchronized (appenders) {
+      appenders.forEach(appender -> safelyExecute(
+          appender::stop,
+          String.format("Failed to stop appender: %s", appender.getName())
+      ));
+    }
+  }
+
+  @Override
+  protected void appendEvent(LogMessage logMessage) {
+    synchronized (appenders) {
+      appenders.forEach(appender -> safelyExecute(
+          () -> appender.appendEvent(logMessage),
+          String.format("Failed to append log message to appender: %s. Message: %s",
+              appender.getName(), logMessage.getFormattedMessage())
+      ));
+    }
+  }
+
+  @Override
+  public void setContext(Context context) {
+    super.setContext(context);
+    synchronized (appenders) {
+      appenders.forEach(appender -> safelyExecute(
+          () -> appender.setContext(context),
+          String.format("Failed to set context to appender: %s", appender.getName())
+      ));
+    }
+  }
+
+  @Override
+  public synchronized void doAppend(ILoggingEvent eventObject) {
+    if (shouldSkipLogging()) {
+      return;
+    }
+    super.doAppend(eventObject);
+  }
+
+  boolean shouldSkipLogging() {
```

Comment:
is this method non-private only for testing?

If yes, please use `@VisibleForTesting` annotation.

Reply:
Done